### ISSUE TITLE
Catalog failure when rerunning a multiple install

### ISSUE THREAD

**Issue by yohanboniface** [ideascube/ideascube#326]
```
Traceback (most recent call last):
  File "/usr/share/python/ideascube/bin/django-admin", line 11, in <module>
    sys.exit(execute_from_command_line())
  File "/usr/share/python/ideascube/lib/python3.4/site-packages/django/core/management/__init__.py", line 354, in execute_from_command_line
    utility.execute()
  File "/usr/share/python/ideascube/lib/python3.4/site-packages/django/core/management/__init__.py", line 346, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/share/python/ideascube/lib/python3.4/site-packages/django/core/management/base.py", line 394, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/share/python/ideascube/lib/python3.4/site-packages/django/core/management/base.py", line 445, in execute
    output = self.handle(*args, **options)
  File "/usr/share/python/ideascube/lib/python3.4/site-packages/ideascube/serveradmin/management/commands/catalog.py", line 105, in handle
    options['func'](options)
  File "/usr/share/python/ideascube/lib/python3.4/site-packages/ideascube/serveradmin/management/commands/catalog.py", line 151, in install_packages
    self.catalog.install_packages(options['ids'])
  File "/usr/share/python/ideascube/lib/python3.4/site-packages/ideascube/serveradmin/catalog.py", line 444, in install_packages
    handler.install(pkg, download_path)
  File "/usr/share/python/ideascube/lib/python3.4/site-packages/ideascube/serveradmin/catalog.py", line 78, in install
    package.install(download_path, self._install_dir)
  File "/usr/share/python/ideascube/lib/python3.4/site-packages/ideascube/serveradmin/catalog.py", line 247, in install
    os.rename(path, new)
OSError: [Errno 39] Directory not empty: '/var/ideascube/kiwix/data/index/ubuntudoc_fr_all_2015-12.zim.idx' -> '/var/ideascube/kiwix/data/index/ubuntudoc.fr.zim.idx'
```

This is after rerunning the command `sudo ideascube catalog install ubuntudoc.fr cest-pas-sorcier.fr wikisource.fr wikibooks.fr wikibooks.en wikiversity.fr wikiversity.en wiktionary.fr wiktionary.en wikipedia.ki wikipedia.f r wikipedia.en` which has previously failed during the download of cest-pas-sorcier. So ubuntudoc.fr has been installed but not yet marked as installed by the catalog.

I think we need to fix two things here:
- mark as installed a package when installed, without waiting for all the other packages to be processed
- do not fail if the idx already exist when installing a zim

**Comment by bochecha** [comment]
> mark as installed a package when installed, without waiting for all the other packages to be processed

Can we?

Installing a package is 3 things:

1. unpacking the files contained in the package
2. running the `commit` method for the handler
3. marking the package as installed

You're suggesting we move step 3 before step 2.

However, installing 2 (or more) packages goes as follows:

1. unpack the files contained in the first package
2. unpack the files contained in the second package
3. run the `commit` method for the handler (let's assume both packages use the same handler)
4. mark the first package as installed
5. mark the second package as installed

If we move steps 4 and 5 before step 3, and if the `commit` method fails... should the packages really be considered installed?

This is not a rhetorical question, both answers are valid, but the decision has an impact on the semantics of the package manager and what it will be able to do.

If we decide that the `commit` method can fail and the package should still be considered as installed, then we need a way to rerun the `commit` method manually.

If we decide that a failure in the `commit` method should mean the package isn't installed, then we need to consider the whole transaction (download(s), unpack(s), commit(s), mark(s)) as atomic, and we need to rollback what was done when a failure happens.

For what is worth, RPM chose the first option. Packages have post-install, post-uninstall,... scripts. Failures in those scripts should never be fatal. If a script fails, the package is still marked as installed/removed/updated/...

If we decide to do that, then instead of changing the order of the steps as described above, I think a better way would be to surround each `commit` by `try: except:` blocks, and just continue to the next `commit` on failures.

We probably also should download everything first, then start installing/committing.

**Comment by bochecha** [comment]
> do not fail if the idx already exist when installing a zim

We probably want to take some inspiration from other package managers here.

RPM just unpacks the rpm package (it's basically a cpio archive), overwriting whatever already existed.

So we probably want to do the same, not bother with anything already installed and just make sure whatever we had to unpack gets unpacked over.

**Comment by yohanboniface** [comment]
> If we decide that the commit method can fail and the package should still be considered as installed, then we need a way to rerun the commit method manually.

This is where the "service management" page should pop up ;)

I take your good points. I'm a bit balanced. I'd say that, in theory, the atomic transaction is the best option, but I feel like in practice this would be quite hard to achieve for now.

> if the commit method fails... should the packages really be considered installed?

I think that if the package.install run without any failure, it's OK to consider it as installed.

> If we decide to do that, then instead of changing the order of the steps as described above, I think a better way would be to surround each commit by try: except: blocks, and just continue to the next commit on failures.

Yes, I'm tempted to catch processing error anyway. But doing so we need a way to do a summary at the end for the user to know which ones have been effectively installed and which one have failed.

> So we probably want to do the same, not bother with anything already installed and just make sure whatever we had to unpack gets unpacked over.

Sounds good to me. :)

I'll update the PR #327 accordingly and ping you again when it's ready for review :)

Thanks for your feed back!

**Comment by bochecha** [comment]
> in theory, the atomic transaction is the best option, but I feel like in practice this would be quite hard to achieve for now.

Not just "*for now*".

Let's once again take a look at what other, more established package managers do...

Rpm tries to do transactions... but it doesn't really. What it does is run everything first as a "dry run", without doing anything, just pretending to do it "as if". If an error happens during this "test transaction", then the "actual transaction" won't be performed at all. This helps detecting things like file conflicts and other errors, but it doesn't (and can't) catch all possible failures. In particular, I **think** that the "test transaction" won't run the post-install/post-uninstall/... scripts. As a result, it is still possible to have the "actual transaction" fail in the middle, although it is quite rare these days (in no small part because each distribution has strong guidelines about how to properly package things, and we try really hard not to break stuff).

As far as I know, the deb/dpkg/apt stack doesn't even try doing transactions at all. Not proper ones, not even "pretend" ones like rpm.

I wouldn't presume that we'd be capable of doing something these communities have failed to do for so many years, but hey, I'd love to be surprised. :smile:

**Comment by yohanboniface** [comment]
Fixed by #327
